[quant][pt1] Add the dynamic quantized LSTM module

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25157 [quant][pt1] Add the dynamic quantized LSTM module**

Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
[WIP][quant][pt1] Add the dynamic quantized LSTM module

Still need to verify the correctness (Currently it outputs the correct results).

TODO:
- Refine the unit test.
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[WIP][quant][pt1] Add the dynamic quantized LSTM module"


Still need to verify the correctness (Currently it outputs the correct results).

TODO:
- Refine the unit test.
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[WIP][quant][pt1] Add the dynamic quantized LSTM module"


Still need to verify the correctness (Currently it outputs the correct results).

TODO:
- Refine the unit test.
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jamesr66a commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -0,0 +1,370 @@
+        # print(type(mod))
+        # assert LSTM or GRU
+        # assert type(mod) == torch.nn.quantized.dynamic.RNNBase.from_float,
+        #     'nn.quantized.dynamic.RNNBase.from_float only works for nn.LSTM'
```
Why is this commented?

jianyuh replied:
Let me uncomment this.

jamesr66a commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
   }
 };
 
+struct QuantizedCellParamsDynamic {
```
Is the purpose of this struct that we do not require explicit scale and zero point parameters? You should add a comment here explaining why we have both `QuantizedCellParams` and `QuantizedCellParamsDynamic`

jianyuh replied:
That is right. Basically the Tensor in the arguments encapsulates the scale and the zero point in the struct in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h#L18-L23. I will add a comment here.

jamesr66a reviewed (commented):
This is a good start!

There's a ton of stuff commented, I'm assuming because some functionality doesn't work. We should add comments and/or issues documenting why these features are missing.

Additionally, I think there's some inconsistency in the code around whether LSTM and GRU or just LSTM is supported. We should clean up the code to ensure that everything is consistent.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -3,10 +3,38 @@
+      c10::Dispatcher::singleton().findSchema({func_name, overload_name});
+  assert(op_handle.has_value());
+  return callOp(op_handle.value(), args...);
+}
```
@smessmer : Do you think if we can put this in C10 itself? Previously I chatted with you in the group, you pointed me to this solution in order to call c10 ops from ATen code (in C++). I took these code pieces from `kernel_functor_test.cpp`.

jianyuh replied:
I now separate the c10 usage functions in an independent file: c10_utils.h. Hope this resolve the issue.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
+struct QuantizedCellParamsDynamic {
+  QuantizedCellParamsDynamic(
+      const Tensor& _w_ih, /* QTensor, qint8 */
+      const Tensor& _w_hh, /* QTensor, qint8 */
```
Right. It is pre-packed. Will update the comment.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
+  template <class... Inputs>
+  inline std::vector<c10::IValue> makeStack(Inputs&&... inputs) {
+    return {std::forward<Inputs>(inputs)...};
+  }
```
Will remove this.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +166,52 @@ struct QuantizedCellParams {
+    const auto kFuncName = "quantized::fbgemm_linear_dynamic";
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_ih_list = callOp(
+        kFuncName, kOvrldName, input_ih, w_ih, b_ih);
```
Sure: will add an assert.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1014,6 +1101,56 @@ std::tuple<Tensor, Tensor, Tensor> quantized_lstm(
+// an int8 quantized weight. This is advantageous in small-batch-size
+// scenarios where runtime is dominated by memory fetches of the weight matrix.
+
+std::tuple<Tensor, Tensor, Tensor> quantized_lstm_dynamic(
```
Right: it is copied from there. Let me check if we can reuse one copy of the code

jianyuh replied:
I will reuse the same `quantized_lstm` function with a new argument `use_dynamic` to check if we should use the new prepacked weight with the `PackedLinearWeight` struct.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -0,0 +1,370 @@
+#     # @torch.jit.script_method
+#     # _load_from_state_dict
+#     # __setstate__
+#     # load from the file
```
The serialization is not implemented in this PR. Maybe let me remove that in this PR and add a new PR to implement the serialization.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -0,0 +1,370 @@
+        assert mod.bias
+
+        # TODO: support more than just LSTM
+        if qRNNBase.mode != 'LSTM' and qRNNBase.mode != 'GRU':
```
I should only keep `if qRNNBase.mode != 'LSTM'` as GRU is not implemented yet.

jianyuh reviewed (commented):
Thanks @jamesr66a ! Will work on the modification for this PR. As mentioned, I will separate the serialization as a new PR.

jianyuh committed:
Update on "[WIP][quant][pt1] Add the dynamic quantized LSTM module"


Still need to verify the correctness (Currently it outputs the correct results).

TODO:
- Refine the unit test.
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[WIP][quant][pt1] Add the dynamic quantized LSTM module"


Still need to verify the correctness (Currently it outputs the correct results).

TODO:
- Refine the unit test.
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[quant][pt1] Add the dynamic quantized LSTM module"


Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[quant][pt1] Add the dynamic quantized LSTM module"


Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh committed:
Update on "[quant][pt1] Add the dynamic quantized LSTM module"


Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jamesr66a reviewed (approved):
LGTM!

jianyuh committed:
Update on "[quant][pt1] Add the dynamic quantized LSTM module"


Add the dynamic quantized LSTM module.

TODO (separate PRs):
- Serialization.
- Bias can be Null.

Differential Revision: [D16955662](https://our.internmc.facebook.com/intern/diff/D16955662/)

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +140,59 @@ struct QuantizedCellParams {
+// QuantizedCellParams vs. QuantizedCellParamsDynamic
+//
+// QuantizedCellParams uses the legacy
+// fbgemm_linear_int8_weighjt_fp32_activation API, which requires the explicit
```
Fixed.

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -138,6 +140,59 @@ struct QuantizedCellParams {
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_ih_list =
+        callOp(kFuncName, kOvrldName, input_ih, w_ih, b_ih);
+    TORCH_CHECK(
```
Fixed.
